wgpu-core/src/weak_vec.rs

```diff
@@ -0,0 +1,70 @@
+//! Module containing the [`WeakVec`] API.
+
+use std::sync::Weak;
+
+/// A container that holds Weak references of T.
+///
+/// On `push` it scans its contents for weak references with no strong references still alive and drops them.
+#[derive(Debug)]
+pub(crate) struct WeakVec<T> {
+    inner: Vec<Option<Weak<T>>>,
+}
+
+impl<T> Default for WeakVec<T> {
+    fn default() -> Self {
+        Self {
+            inner: Default::default(),
+        }
+    }
+}
+
+impl<T> WeakVec<T> {
+    pub(crate) fn new() -> Self {
+        Self { inner: Vec::new() }
+    }
+
+    /// Pushes a new element to this collection, dropping older elements that no longer have
+    /// a strong reference to them.
+    ///
+    /// NOTE: The length and capacity of this collection do not change when old elements are
+    /// dropped.
+    pub(crate) fn push(&mut self, value: Weak<T>) {
```

Comment:
nitpick: This comment about what the function does would really be better served as a doc. comment.

Pushed as a `fixup!` with ca93e72b.